knora/dsplib/utils/excel_to_json_lists.py

```diff
@@ -1,145 +1,176 @@
-"""This module handles all the operations which are used for the creation of JSON lists form Excel files."""
+"""This module handles all the operations which are used for the creation of JSON lists from Excel files."""
 import csv
 import glob
 import json
 import os
 import re
 import unicodedata
+from typing import Any, Union, Optional
 
 import jsonschema
-from jsonschema import validate
 from openpyxl import load_workbook
-
-list_of_lists = []
-cell_names = []
-
-
-def get_values_from_excel(excelfiles: list[str], base_file: str, parentnode: {}, row: int, col: int,
-                          preval: list[str]) -> int:
+from openpyxl.cell import Cell
+from openpyxl.worksheet.worksheet import Worksheet
+
+# Global variables used to ensure that there are no duplicate node names
+list_of_lists_of_previous_cell_values: list[list[str]] = []
+list_of_previous_node_names: list[str] = []
+
+
+def get_values_from_excel(
+    excelfiles: dict[str, Worksheet],
+    base_file: dict[str, Worksheet],
+    parentnode: dict[str, Any],
+    row: int,
+    col: int,
+    preval: list[str]
+) -> tuple[int, dict[str, Any]]:
     """
-    This function calls itself recursively to go through the Excel files. It extracts the cell values and creates the JSON list
-    file.
+    This function calls itself recursively to go through the Excel files. It extracts the cell values and creates
+    the JSON list file.
 
     Args:
-        base_file: File name of the base file
         excelfiles: List of Excel files with the values in different languages
-        parentnode: Name(s) of the parent node(s) of the actual node
-        row: The index of the actual row of the Excel sheet
-        col: The index of the actual column of the Excel sheet
+        base_file: File name of the base file
+        parentnode: Name(s) of the parent node(s) of the current node
+        row: The index of the current row of the Excel sheet
+        col: The index of the current column of the Excel sheet
         preval: List of previous values, needed to check the consistency of the list hierarchy
 
     Returns:
-        int: Row index for the next loop (actual row index minus 1)
+        int: Row index for the next loop (current row index minus 1)
+        dict: The JSON list up to the current recursion. At the last recursion, this is the final JSON list.
     """
-    nodes = []
-    currentnode = {}
-    wb = load_workbook(filename=base_file, read_only=True)
-    worksheet = wb.worksheets[0]
-    cell = worksheet.cell(column=col, row=row)
+    nodes: list[dict[str, Any]] = []
+    currentnode: dict[str, Any] = dict()
+    base_file_ws: Worksheet = list(base_file.values())[0]
+    cell: Cell = base_file_ws.cell(column=col, row=row)
+
+    for excelfile in excelfiles.values():
+        if any((not excelfile['A1'].value, excelfile['B1'].value)):
```

Comment:
I think it is a good idea to check if A1 has a value, but it seems somewhat random to check if there is also a vaue in B1 (what about the other cells in this row?). I don't know if it is easily possible to check if there is a value in any other cell besides A1. But that would certainly be more elegant.

Reply:
I see your point, but I think there is no alternative, because in openpyxl the row goes on until the maximum allowed in Excel, which is a lot. (In pandas, the row goes only as far as the longest row goes, so it is easy to check if all cells except the first are empty). 
In practice, I think that if B1 is empty, it is almost granted that the A1 is the only value in this row.

Reply:
OK, if it is not easily possible to check for all the other cells in the row, we don't need to do it. Your solution is fine than.